ecommers/lib/ui/pages/authorization/sign_up_page.dart

```diff
@@ -11,66 +10,134 @@ import 'package:ecommers/ui/widgets/button/index.dart';
 import 'package:provider/provider.dart';
 
-class SignUpPage extends StatelessWidget {
+class SignUpPage extends StatefulWidget {
+  @override
+  _SignUpPageState createState() => _SignUpPageState();
+}
+
+class _SignUpPageState extends State<SignUpPage> {
+  TextEditingController phoneController = TextEditingController();
+  TextEditingController emailController = TextEditingController();
+  TextEditingController passwordController = TextEditingController();
+  TextEditingController userNameController = TextEditingController();
+  bool _isPhoneSelected = false;
+
+  @override
+  void dispose() {
+    phoneController.dispose();
+    emailController.dispose();
+    passwordController.dispose();
+    userNameController.dispose();
+    super.dispose();
+  }
+
   @override
   Widget build(BuildContext context) {
-    final localization = I18n.of(context);
     final provider = Provider.of<SignUpProviderModel>(context, listen: false);
 
-    final signUpForm = _buildSignUpForm(localization, provider);
+    final authForm = _isPhoneSelected
+        ? _buildPhoneSignUp(provider)
+        : _buildSignUpForm(provider);
 
     return AuthorizationTabBase(
       children: <Widget>[
         const SizedBox(height: Insets.x5),
-        signUpForm,
+        authForm,
         const SizedBox(height: Insets.x3_5),
         PrimaryButtonWidget(
           text: localization.signUp,
           assetIconPath: Assets.arrowRightIcon,
-          onPressedFunction: () {
-            if (signUpForm.formKey.currentState.validate()) {
-              provider.tryAuthorize();
-            }
-          },
+          onPressedFunction: () => _onSignUpTapped(authForm, provider),
+        ),
+        FlatButton(
+          onPressed: () => setState(() => _isPhoneSelected = !_isPhoneSelected),
+          child: Text(
+            _isPhoneSelected
+                ? localization.or_use_email
+                : localization.or_use_phone,
+            style: Theme.of(context)
+                .textTheme
+                .bodyText1
+                .copyWith(color: BrandingColors.primary),
+          ),
         ),
         const SizedBox(height: Insets.x8_5),
         AuthRichText(textSpanModelList: provider.bottomText),
       ],
     );
   }
 
-  AuthForm _buildSignUpForm(
-      I18n localization, SignUpProviderModel provider) {
+  AuthForm _buildPhoneSignUp(SignUpProviderModel provider) {
+    return AuthForm(
+      child: Column(
+        mainAxisSize: MainAxisSize.min,
+        children: <Widget>[
+          AuthTextField(
+            labelText: localization.your_phone_number,
+            keyboardType: TextInputType.phone,
+            icon: Icons.phone,
+            prefixText: '+',
+            controller: phoneController,
+            onValidate: (text) => UserValidator.isPhoneNumber(text)
+                ? null
+                : localization.incorrect_phone_number,
+            onChanged: (text) => provider.phoneNumber = phoneController.text,
```

Comment:
do we really need to use controller? because if we use it just to set providers field we can use (text) argument that comes to us.

Reply:
Sure, for example if you enter text and then swipe to left or right, also with hot reload text saves(easy debug)